core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1984,6 +1984,42 @@ object Stream {
       self.reduce(S.combine(_, _))
 
     /**
+      * Repartitions the input with the function `p`. On each step `p` is applied
+      * to the input and all elements but the last of the resulting sequence
+      * are emitted. The last element is then appended to the next input using the
+      * Semigroup `S`.
+      *
+      * @example {{{
+      * scala> import cats.implicits._
+      * scala> Stream("Hel", "l", "o Wor", "ld").repartition(s => Segment.array(s.split(" "))).toList
+      * res0: List[String] = List(Hello, World)
+      * }}}
+      */
+    def repartition(p: O => Segment[O, Unit])(implicit S: Semigroup[O]): Stream[F,O] = {
+      def go(carry: Option[O], s: Stream[F,O]): Pull[F,O,Unit] = {
+        s.pull.uncons.flatMap {
+          case Some((ohd, otl)) =>
+            ohd.uncons1 match {
+              case Right((hhd,htl)) =>
+                val tl = Stream.segment(htl) ++ otl
+                val next = carry.fold(hhd)(c => S.combine(c, hhd))
+                val parts = p(next).toVector
```

Comment:
We should avoid using `toVector` here (e.g., b/c the segment could be infinite). One option is changing the signature of `repartition` to `def repartition(p: O => Chunk[O])(implicit S: Semigroup[O]): Stream[F,O]`. I'd be better to keep the current signature and use one of the lazy/fused combinators on `Segment` though. Let me know how it goes or if you get stuck -- I think we should be able to get it to work.

Reply:
Done, with a test case with an infinite segment. I had some trouble getting this to work with `uncons` instead of `uncons1` because `repartition` applies `p` to `carry ++ head`, and needs to emit all partitions before determining the next `carry`.

Do you have any suggestions on how to avoid `uncons1`? Thanks!

Reply:
General idea is to split the functionality of `repartition` up in to 2 layers -- an outer layer which does `s.pull.uncons.flatMap(...)` and an inner layer which statefully scans each segment. The segment scan state would be the current carry value. Each step of the scan would output an `Option[O]`.

Roughly, this would look something like:

```scala
def repartition(p: O => Segment[O,Unit])(implicit S: Semigroup[O]): Stream[F,O] = {
  def process(carry: Option[O], s: Segment[O,Unit]): Segment[O,Option[O]] = {
    s.scan(carry)((carry, o) => /* TODO compute an output value using carry and o */).collect { case Some(o) => o }
  }
  def go(carry: Option[O], s: Stream[F,O]): Pull[F,O,Unit] = {
    s.pull.uncons.flatMap {
      case None => carry.map(Pull.output1(_)).getOrElse(Pull.done)
      case Some((hd,tl)) =>
        Pull.segment(process(carry, hd)).flatMap { newCarry => go(newCarry, tl) }
    }
  }
  go(None, this)
}
```

Reply:
Right now, what we do is effectively: given an input element and a carry, partition `carry |+| input`, and emit `partitions.init`; use `partitions.last` as the carry when the next input comes in.

In the body of the scan, how do we emit all finished partitions (`partitions.init`) while threading through the carry (`partitions.last`) for each element? Currently this is done by recursively returning pulls to output finished partitions until we get to the end, where we make a recursive call with a carry (the unfinished partition). I'm not sure how to do this with `Segment`.